File: ParallelAndVectorized/Algorithms/FixedThreadsMatrixMultiplication.py
from concurrent.futures import ThreadPoolExecutor
import numpy as np

class FixedThreadsMatrixMultiplication:
    def multiply(self, matrixA, matrixB, num_threads):
        size = len(matrixA)
        result = [[0.0 for _ in range(size)] for _ in range(size)]

        def compute_row(row):
            for j in range(size):
                for k in range(size):
                    result[row][j] += matrixA[row][k] * matrixB[k][j]

        with ThreadPoolExecutor(max_workers=num_threads) as executor:
            executor.map(compute_row, range(size))

        return np.array(result)

File: ParalellAndVectorizedMatrices/src/main/java/org/example/FixedThreadsMatrixMultiplication.java
package org.example;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

public class FixedThreadsMatrixMultiplication implements ParallelMatrixMultiplication{

    @Override
    public double[][] multiply(double[][] matrixA, double[][] matrixB, int numThreads) throws InterruptedException {
        int size = matrixA.length;
        double[][] result = new double[size][size];

        ExecutorService executorService = Executors.newFixedThreadPool(numThreads);

        for (int i = 0; i < size; i++) {
            final int row = i;
            executorService.submit(() -> {
                for (int j = 0; j < size; j++) {
                    for (int k = 0; k < size; k++) {
                        result[row][j] += matrixA[row][k] * matrixB[k][j];
                    }
                }
            });
        }

        executorService.shutdown();
        executorService.awaitTermination(1, TimeUnit.HOURS);

        return result;
    }
}
